### PR TITLE
Show white spaces in Diff Viewer

### LINKED ISSUE #202: User Feedback: Whitespace
Based on what I have read from the HSS surveys, it seems like a lot of people are having problems with the whitespaces/line skips. People will have the correct output but in a slightly different format and at times they will get extremely little to no credit. We should look into finding a way to check if the output is correct when looking past the formatting differences.

### PR DESCRIPTION
**Pull request by saxocellphone**
Now the diff viewer shows some defined visuals of empty spaces. A few things to note:

- Some empty chars like backspace just simply remove a character, so the diffviewer can't register it as a special character. 

- I added an extra function in diffviewer that makes it easy to add new visualizations in the future. 

- I made the line number line up better. Say that there are 200 lines, every line number is going to get 3 characters. This way, the display doesn't get shifted when switching decimal places.

If I remember correctly, this should close #202 and part of #802

### CODE REVIEW

File: site/app/libraries/DiffViewer.php [hunk 1]
```diff
@@ -415,24 +420,64 @@ private function getDisplay($lines, $type="expected") {
             else {
                 $html .= "\t<div>";
             }
-            $html .= "<span class='line_number'>{$j}</span>";
+            $html .= "<span class='line_number'>";
+            $digits_at_line = strlen((string)$j);
+			$max_digits = strlen((string)count($lines));
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]: fix whitespace

File: site/app/libraries/DiffViewer.php [hunk 2]
```diff
@@ -415,24 +420,64 @@ private function getDisplay($lines, $type="expected") {
...
+            $counter = $max_digits-$digits_at_line;
+            while($counter > 0){
+				$counter--;
+				$html .= "&nbsp;";
+			}
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]:
```
for ($counter = ($max_digits - $digits_at_line); $counter > 0; $counter--) {
    $html .= "&nbsp;";
}
```

File: site/app/libraries/DiffViewer.php [hunk 3]
```diff
@@ -415,24 +420,64 @@ private function getDisplay($lines, $type="expected") {
...
+						$html_no_empty = $this->replaceEmptyChar($html_orig);
+						$html_no_empty_error = $this->replaceEmptyChar($html_orig_error);
+						$html .= $html_no_empty;
+						$html .= "<span class='highlight-char'>".$html_no_empty_error."</span>";
+					} else if($option == "with_escape") {
+						$html_no_empty = $this->replaceEmptyCharWEscape($html_orig);
+						$html_no_empty_error = $this->replaceEmptyCharWEscape($html_orig_error);
+						$html .= $html_no_empty;
+						$html .= "<span class='highlight-char'>".$html_no_empty_error."</span>";
+					}
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]: fix identation

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 4]
```diff
@@ -59,13 +59,48 @@ public function run() {
...
 
+    public function ajaxRemoveEmpty(){
+        //This function shows the empty spaces in the diffViewer
+        //TODO: Need to add checks?
+        $gradeable_id = $_REQUEST['gradeable_id'];
+        $gradeable = $this->core->getQueries()->getGradeable($gradeable_id, $_REQUEST['who_id']);
+        $gradeable->loadResultDetails();
+        $testcase = $gradeable->getTestcases()[$_REQUEST['index']];
+        //There are three options: original (Don't show empty space), escape (with escape codes), and unicode (with characters)
+        $option = isset($_REQUEST['option']) ? $_REQUEST['option'] : 'original';
```
Comment (MasterOdin) [inline site/app/controllers/grading/ElectronicGraderController.php]: `$option = $_REQUEST['option'] ?? 'original';`

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 5]
```diff
@@ -59,13 +59,48 @@ public function run() {
...
+        //There are currently two views, the view of student's code and the expected view.
+        $which = isset($_REQUEST['which']) ? $_REQUEST['which'] : 'actual';
```
Comment (MasterOdin) [inline site/app/controllers/grading/ElectronicGraderController.php]: `$which = $_REQUEST['which'] ?? 'actual';`

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 6]
```diff
@@ -59,13 +59,48 @@ public function run() {
...
+        $this->core->getOutput()->useHeader(false);
+        $html = "";
+        if($which == "actual"){
+            $html .= $diff_viewer->getDisplayActual($option);
+        }
+        else {
+            $html .= $diff_viewer->getDisplayExpected($option);
+        }
+        $white_spaces = $diff_viewer->getWhiteSpaces();
+        echo json_encode(array("html"=>"$html", "whitespaces"=>$white_spaces));
```
Comment (MasterOdin) [inline site/app/controllers/grading/ElectronicGraderController.php]: you can get rid of lines 90-91 and the echo and just do:
`$this->core->renderJson(['html' => $html, 'whitespaces' => $white_spaces]);`

File: site/app/libraries/DiffViewer.php [hunk 7]
```diff
@@ -398,13 +404,12 @@ private function getDisplay($lines, $type="expected") {
...
             }
         }
-
         /*
          * Run through every line, starting a highlight around any group of mismatched lines that exist (whether
          * there's a difference on that line or that the line doesn't exist.
          */
         for ($i = 0; $i < count($lines); $i++) {
-            $j = $i + 1;
+        	$j = $i + 1;
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]: You really should configure your editor to not use tabs. There's a couple other instances where you've used tabs in this file.

File: site/app/libraries/DiffViewer.php [hunk 8]
```diff
@@ -415,24 +420,63 @@ private function getDisplay($lines, $type="expected") {
             else {
                 $html .= "\t<div>";
             }
-            $html .= "<span class='line_number'>{$j}</span>";
+            $html .= "<span class='line_number'>";
+            $digits_at_line = strlen((string)$j);
+            $max_digits = strlen((string)count($lines));
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]: this line should be outside the for loop as it doesn't change

File: site/app/libraries/DiffViewer.php [hunk 9]
```diff
@@ -415,24 +420,63 @@ private function getDisplay($lines, $type="expected") {
...
+                		$html_no_empty = $this->replaceEmptyChar($html_orig);
+						$html_no_empty_error = $this->replaceEmptyChar($html_orig_error);
+						$html .= $html_no_empty;
+						$html .= "<span class='highlight-char'>".$html_no_empty_error."</span>";
+					} else if($option == "with_escape") {
+						$html_no_empty = $this->replaceEmptyCharWEscape($html_orig);
+						$html_no_empty_error = $this->replaceEmptyCharWEscape($html_orig_error);
+						$html .= $html_no_empty;
+						$html .= "<span class='highlight-char'>".$html_no_empty_error."</span>";
+					}
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]: fix indentation to not use tabs, but spaces

File: site/app/libraries/DiffViewer.php [hunk 10]
```diff
@@ -415,24 +420,63 @@ private function getDisplay($lines, $type="expected") {
...
+                		$html .= $html_orig;
+                		$html .= "<span class='highlight-char'>".$html_orig_error."</span>";
+					} else if($option == "with_unicode") {
+                		$html_no_empty = $this->replaceEmptyChar($html_orig);
+						$html_no_empty_error = $this->replaceEmptyChar($html_orig_error);
+						$html .= $html_no_empty;
+						$html .= "<span class='highlight-char'>".$html_no_empty_error."</span>";
+					} else if($option == "with_escape") {
+						$html_no_empty = $this->replaceEmptyCharWEscape($html_orig);
+						$html_no_empty_error = $this->replaceEmptyCharWEscape($html_orig_error);
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]: need to update this function reference to the new name?

File: site/app/libraries/DiffViewer.php [hunk 11]
```diff
@@ -452,11 +496,56 @@ private function getDisplay($lines, $type="expected") {
...
+		return $return;
+	}
+
+	private function replaceUTF(&$text, $which, $what, $description){
+		$count = 0;
+		$what = '<span style="outline:1px blue solid;">'.$what.'</span>';
+		$return = str_replace($which, $what, $text,$count);
+		if($count > 0) $this->white_spaces[$description] = $what;
+		return $return;
+	}
```
Comment (MasterOdin) [inline site/app/libraries/DiffViewer.php]: unused function?

**Comment by bmcutler** [comment]
screen shot 2018-06-26 at 4 05 28 pm

**Comment by saxocellphone** [comment]
Note: also add space between legend and viewer

**Comment by bmcutler** [comment]
Why are the newline characters different in the legend and the file display?

screen shot 2018-06-27 at 10 59 19 am

I think you need to use the same (fixed-width) font in the legend as the diff viewer?

Also, the legend is still squished up very close to the title.  Can you please add more vertical space between the title & the legend?  (And it could also be side by side with the "Display whitespace ... " button

**Comment by saxocellphone** [comment]
I'm not really sure what's going on but the behavior in the picture does not happen on my machine. I added a little more space, and made the font the same as the line_code

**Comment by bmcutler** [comment]
Yes, the font is fixed, but you have not increased the space between the title & the legend.
Instead you've added alot of space below the title/legend (that's always there, even if there is no legend).
screen shot 2018-06-27 at 11 53 04 am
screen shot 2018-06-27 at 11 53 13 am

**Comment by bmcutler** [comment]
Going to merge this now, even though the alignment problem is unfixed (only appears on Mac Chrome?)

**Pull request merged**